<a href="https://colab.research.google.com/github/PravallikaaMohan/Machine_Learning/blob/main/Loan_Approval_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Prediction of Loan Application Approval**

We are going to predict the probaility of a loan application being approved.

Dataset:https://raw.githubusercontent.com/callxpert/datasets/master/Loan-applicant-details.csv

### **Loading the Dataset**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import model_selection
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier, export_graphviz

In [ ]:
url = "https://raw.githubusercontent.com/callxpert/datasets/master/Loan-applicant-details.csv"
names = ['Loan_ID','Gender','Married','Dependents','Education','Self_Employed','ApplicantIncome','CoapplicantIncome','LoanAmount','Loan_Amount_Term','Credit_History','Property_Area','Loan_Status']
dataset = pd.read_csv(url, names=names)

In [ ]:
dataset.shape

(480, 13)

In [ ]:
print(dataset.head(20))

     Loan_ID  Gender Married  ... Credit_History Property_Area Loan_Status
0   LP001003    Male     Yes  ...              1         Rural           N
1   LP001005    Male     Yes  ...              1         Urban           Y
2   LP001006    Male     Yes  ...              1         Urban           Y
3   LP001008    Male      No  ...              1         Urban           Y
4   LP001011    Male     Yes  ...              1         Urban           Y
5   LP001013    Male     Yes  ...              1         Urban           Y
6   LP001014    Male     Yes  ...              0     Semiurban           N
7   LP001018    Male     Yes  ...              1         Urban           Y
8   LP001020    Male     Yes  ...              1     Semiurban           N
9   LP001024    Male     Yes  ...              1         Urban           Y
10  LP001028    Male     Yes  ...              1         Urban           Y
11  LP001029    Male      No  ...              1         Rural           N
12  LP001030    Male     

In [ ]:
from sklearn.preprocessing import LabelEncoder
var_mod = ['Gender','Married','Dependents','Education','Self_Employed','Property_Area','Loan_Status']
le = LabelEncoder()
for i in var_mod:
    dataset[i] = le.fit_transform(dataset[i])

### **Splitting the Dataset**

Our dataset has 480 records. We are going to use 80% of it for training the model and 20% of the records to evaluate our model. 

Though our dataset has lot of columns, we are only going to use the Income fields, loan amount, loan duration and credit history fields to train our model.

In [ ]:
array = dataset.values
X = array[:,6:11]
Y = array[:,12]
x_train, x_test, y_train, y_test = model_selection.train_test_split(X, Y, test_size=0.2, random_state=7)

In [ ]:
x_train

array([[3522, 0.0, 81, 180, 1],
       [6000, 0.0, 141, 360, 1],
       [3333, 0.0, 70, 360, 1],
       ...,
       [3500, 1667.0, 114, 360, 1],
       [3427, 0.0, 138, 360, 1],
       [6250, 0.0, 128, 360, 1]], dtype=object)

In [ ]:
y_train

array([0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
       0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1,

In [ ]:
x_test

array([[2281, 0.0, 113, 360, 1],
       [6125, 1625.0, 187, 480, 1],
       [1782, 2232.0, 107, 360, 1],
       [1958, 2436.0, 131, 360, 1],
       [2333, 2417.0, 136, 360, 1],
       [6000, 0.0, 205, 240, 1],
       [39999, 0.0, 600, 180, 0],
       [4600, 0.0, 73, 180, 1],
       [2785, 2016.0, 110, 360, 1],
       [4950, 0.0, 125, 360, 1],
       [3069, 0.0, 71, 480, 1],
       [7583, 0.0, 187, 360, 1],
       [7948, 7166.0, 480, 360, 1],
       [2882, 1843.0, 123, 480, 1],
       [5500, 0.0, 105, 360, 0],
       [5568, 2142.0, 175, 360, 1],
       [9963, 0.0, 180, 360, 1],
       [9328, 0.0, 188, 180, 1],
       [10408, 0.0, 259, 360, 1],
       [8072, 240.0, 253, 360, 1],
       [2491, 2054.0, 104, 360, 1],
       [2500, 0.0, 96, 480, 1],
       [3510, 4416.0, 243, 360, 1],
       [16120, 0.0, 260, 360, 1],
       [5250, 0.0, 94, 360, 1],
       [1809, 1868.0, 90, 360, 1],
       [2435, 0.0, 75, 360, 1],
       [2423, 505.0, 130, 360, 1],
       [2192, 1742.0, 45, 360, 1],
       

### **Evaluating and Training the Model**

Logistic Regression : Logistic Regression is a classification algorithm. It is used to predict a binary outcome (1 / 0, Yes / No, True / False) given a set of independent variables. To represent binary / categorical outcome, we use dummy variables


In [ ]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train,y_train)
predictions = model.predict(x_test)
print(accuracy_score(list(y_test), predictions))

0.7708333333333334


Decision tree : Decision tree is a type of supervised learning algorithm (having a pre-defined target variable) that is mostly used in classification problems. It works for both categorical and continuous input and output variables

In [ ]:
model = DecisionTreeClassifier()
model.fit(x_train,y_train)
predictions = model.predict(x_test)
print(accuracy_score(list(y_test), predictions))

0.6354166666666666


Random forest : Random forests or random decision forests are an ensemble learning method for classification, regression and other tasks, that operate by constructing a multitude of decision trees at training time and outputting the class that is the mode of the classes (classification) or mean prediction (regression) of the individual trees

In [ ]:
model = RandomForestClassifier(n_estimators=100)
model.fit(x_train,y_train)
predictions = model.predict(x_test)
print(accuracy_score(list(y_test), predictions))

0.75


So,Logistic Regression Model gives us a better accuracy than other two algorithms.

### **Testing the Algorithm**

Let us check the accuracy for the following data:

0   LP001003    Male     Yes  ...              1         Rural           N

In [ ]:
X = array[0:1,6:11]
Y = array[0:1,12]
X

NameError: ignored

In [ ]:
model = LogisticRegression()
model.fit(x_train,y_train)
predictions = model.predict(X)
print(accuracy_score(list(Y), predictions))

0.0
